src/models/linked_cell/CellGrid.cpp

```diff
@@ -0,0 +1,154 @@
+    : domainSize(domainSize)
+    , cutoffRadius(cutoffRadius)
+    , gridDimensions({ 0, 0, 0 })
+{
+    initializeGrid();
+}
+
+CellGrid::~CellGrid() = default;
+
+void CellGrid::initializeGrid()
+{
+    for (int i = 0; i < 3; ++i) {
+        // Extra cells for halo
+        gridDimensions[i] = static_cast<size_t>(std::ceil(domainSize[i] / cutoffRadius)) + 2;
+    }
+
+    cells.resize(gridDimensions[0]);
+    for (size_t x = 0; x < gridDimensions[0]; ++x) {
+        cells[x].resize(gridDimensions[1]);
+        for (size_t y = 0; y < gridDimensions[1]; ++y) {
+            cells[x][y].resize(gridDimensions[2]);
+            for (size_t z = 0; z < gridDimensions[2]; ++z) {
+                CellType type = determineCellType({ x, y, z });
+                // Initialize vectors for boundary and halo cells
+                std::unique_ptr<Cell> cellPointer;
+                if (type == CellType::Boundary) {
+                    cellPointer = std::make_unique<BoundaryCell>(BoundaryCell({ x, y, z }));
+                    boundaryCells.push_back({ x, y, z });
+                } else if (type == CellType::Halo) {
+                    cellPointer = std::make_unique<HaloCell>(HaloCell({ x, y, z }));
+                    haloCells.push_back({ x, y, z });
+                } else {
+                    cellPointer = std::make_unique<InnerCell>(InnerCell());
+                }
+                cells[x][y][z] = std::move(cellPointer);
+            }
+        }
+    }
+}
+
+CellType CellGrid::determineCellType(const std::array<size_t, 3>& indices) const
+{
+    for (int i = 0; i < 3; ++i) {
+        if (indices[i] == 0 || indices[i] == gridDimensions[i] - 1) {
+            return CellType::Halo;
+        }
+    }
+    for (int i = 0; i < 3; ++i) {
+        if (indices[i] == 1 || indices[i] == gridDimensions[i] - 2) {
+            return CellType::Boundary;
+        }
+    }
+    return CellType::Inner;
+}
+
+// Methods to get boundary and halo particle iterators
+CellGrid::BoundaryIterator CellGrid::beginBoundaryParticles()
+{
+    return BoundaryIterator(boundaryCells, false);
+}
+
+CellGrid::BoundaryIterator CellGrid::endBoundaryParticles()
+{
+    return BoundaryIterator(boundaryCells, true);
+}
+
+CellGrid::HaloIterator CellGrid::beginHaloParticles()
+{
+    return HaloIterator(haloCells, false);
+}
+
+CellGrid::HaloIterator CellGrid::endHaloParticles()
+{
+    return HaloIterator(haloCells, true);
+}
+
+void CellGrid::addParticle(Particle& particle)
+{
+    auto pos = particle.getX();
+    std::array<size_t, 3> indices { 0, 0, 0 };
+
+    for (int i = 0; i < 3; ++i) {
+        if (pos[i] < 0) {
+            indices[i] = 0;
+        } else if (pos[i] >= domainSize[i]) {
+            indices[i] = gridDimensions[i] - 1;
+        } else {
+            indices[i] = static_cast<size_t>(pos[i] / cutoffRadius) + 1;
+        }
+    }
+
+    cells[indices[0]][indices[1]][indices[2]]->addParticle(particle);
+}
```

Comment:
Hier scheinst du davon auszugehen, dass die Cell am Index 0, 0, 0 der Ursprung unseres Koordinatensystem liegt. Aber ich hätte gesagt, dass unser Grid an einer beliebigen Stelle liegen kann und dementsprechend die Partikel auch beliebige Positionen haben kann. D.h. hätte ich gesagt wir speichen einfach in `CellGrid` eine Position, die den Punkt "ganz links oben" von unserer Domäne darstellt. Wenn wir dann das von der Position von den Partiklen abziehen, funktioniert das dann wieder :)
Dabei hätte ich aber gesagt, dass dieser "Ursprungspunkt" unserer Domäne den Anfang der ersten `BoundaryCell` darstellt, weil die `HaloCells` sind ja außerhalb unserer betrachtenden Domäne.

Reply:
Ah ja da hast du recht! Dass der Ursprungspunkt der Domäne bei den BoundaryCells anfängt macht auch Sinn.